server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java

```diff
@@ -129,11 +138,29 @@ private void sendUnpromotableRequests(IndexShard indexShard, long generation, bo
             transportService.getLocalNode(),
             TransportUnpromotableShardRefreshAction.NAME,
             unpromotableReplicaRequest,
-            new ActionListenerResponseHandler<>(
-                listener.delegateFailure((l, r) -> l.onResponse(wasForced)),
-                (in) -> ActionResponse.Empty.INSTANCE,
-                ThreadPool.Names.REFRESH
-            )
+            new ActionListenerResponseHandler<>(ActionListener.wrap(r -> listener.onResponse(wasForced), e -> {
+                ShardRouting shardRouting = indexShard.routingEntry();
+                shardStateAction.remoteShardFailed(
+                    shardRouting.shardId(),
+                    shardRouting.allocationId().getId(),
+                    indexShard.getOperationPrimaryTerm(),
+                    true,
+                    "mark unpromotable copy as stale after refresh failure",
+                    e,
+                    new ActionListener<>() {
+                        @Override
+                        public void onResponse(Void unused) {
+                            listener.onFailure(e);
```

Comment:
This is slightly secondary, because we need to resolve the correct shards being failed. But, I don't think if the shard does not refresh BUT it is failed that we want to error this listener. Erroring this listener fails the entire bulk request I believe.

We only want to error this listener if the refresh fails and we cannot successfully fail the hsard.